src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs

```diff
@@ -0,0 +1,1188 @@
+        {
+            if (value is not KeyValuePair<TKey, TValue> pair)
+            {
+                throw new ArgumentException(SR.Format(SR.Arg_WrongType, value, typeof(KeyValuePair<TKey, TValue>)), nameof(value));
+            }
+
+            Insert(index, pair.Key, pair.Value);
+        }
+
+        /// <inheritdoc/>
+        void IList.Remove(object? value)
+        {
+            if (value is KeyValuePair<TKey, TValue> pair)
+            {
+                ((ICollection<KeyValuePair<TKey, TValue>>)this).Remove(pair);
+            }
+        }
+
+        /// <summary>Provides debug validation of the consistency of the collection.</summary>
+        [Conditional("DEBUG")]
+        private void AssertInvariants()
+        {
+            Debug.Assert(_dictionary.Count == _list.Count, $"Expected dictionary count {_dictionary.Count} to equal list count {_list.Count}");
+            foreach (TKey key in _list)
+            {
+                Debug.Assert(_dictionary.ContainsKey(key), $"Expected dictionary to contain key {key}");
+            }
+        }
+
+        /// <summary>Enumerates the elements of a <see cref="OrderedDictionary{TKey, TValue}"/>.</summary>
+        public struct Enumerator : IEnumerator<KeyValuePair<TKey, TValue>>, IDictionaryEnumerator
+        {
+            /// <summary>The dictionary being enumerated.</summary>
+            private readonly OrderedDictionary<TKey, TValue> _dictionary;
+            /// <summary>The wrapped ordered enumerator.</summary>
+            private List<TKey>.Enumerator _keyEnumerator;
+            /// <summary>Whether Current should be a DictionaryEntry.</summary>
+            private bool _useDictionaryEntry;
+
+            /// <summary>Initialize the enumerator.</summary>
+            internal Enumerator(OrderedDictionary<TKey, TValue> dictionary, bool useDictionaryEntry)
+            {
+                _dictionary = dictionary;
+                _keyEnumerator = dictionary._list.GetEnumerator();
+                _useDictionaryEntry = useDictionaryEntry;
+            }
+
+            /// <inheritdoc/>
+            public KeyValuePair<TKey, TValue> Current { get; private set; }
+
+            /// <inheritdoc/>
+            readonly object IEnumerator.Current => _useDictionaryEntry ?
+                new DictionaryEntry(Current.Key, Current.Value) :
+                Current;
+
+            /// <inheritdoc/>
+            readonly DictionaryEntry IDictionaryEnumerator.Entry => new(Current.Key, Current.Value);
+
+            /// <inheritdoc/>
+            readonly object IDictionaryEnumerator.Key => Current.Key;
+
+            /// <inheritdoc/>
+            readonly object? IDictionaryEnumerator.Value => Current.Value;
+
+            /// <inheritdoc/>
+            public bool MoveNext()
+            {
+                if (_keyEnumerator.MoveNext())
+                {
+                    Current = new(_keyEnumerator.Current, _dictionary._dictionary[_keyEnumerator.Current]);
```

Comment:
In particular, I find this bit to be problematic. Enumerating through the dictionary results in a full lookup of every key that we know is present in the collection. I'm pretty sure that would materialize as a performance regression if I were to use this in STJ property serialization.

Reply:
I'm also a bit concerned about what the performance would end up looking like here as I do quite a lot of enumeration over the key-value pairs. I currently use @TylerBrinkley's implementation [which just indexes an array](https://github.com/dotnet/corefxlab/blob/e0c8705741ad2b66a1bbc70ff5bf5e02886eca1e/src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs#L918).

Reply:
Ok. If we believe enumeration is really the top consideration, I'll redo it.

Reply:
>> If we're really concerned about this, we should scrap using a Dictionary/List separately and do the much more complex but feasible thing of open-coding the whole data structure.

> I don't think we have to go that far.

What else would you do? If enumeration is really the top consideration, then its perf shouldn't fall off a cliff after 9 elements.

Reply:
Whether we make the dictionary optional for small counts is orthogonal to the enumeration concern.

```C#
private readonly Dictionary<TKey, int> _dictionary;
private readonly List<KeyValuePair<TKey, TValue>> _entries;
```

But if we wanted to make it optional this would result in `_dictionary` being created lazily and only being used in key-based lookup. For enumeration the list would always remain the source of truth.

Reply:
> But if we wanted to make it optional this would result in _dictionary being created lazily and only being used in key-based lookup.

Yes, but then inserting / removing requires a lookup for each key in order to update the index; Dictionary doesn't expose anything that lets you update a value for a key as part of an enumeration.

Reply:
> Good point. I guess duplicating the value in both underlying stores is another trade-off?

If the dictionary's value is `(int, TValue)`, then you still have O(N) lookups on remove / insert.

If the dictionary's value is just `TValue`, then `this[key] = value` is an O(N) operation to keep the list's copy in sync.

Reply:
It appears that @TylerBrinkley's implementation also incurs O(n) cost for update operations. So it seems we have to make a choice between optimizing for reads and optimizing for writes?

Reply:
> also incurs O(n) cost for update operations.

So does this one, since `List<T>` needs to do a memory shift. It's not a question of the algorithmic complexity but of the constant factor; are you doing an optimized memcpy, or are you walking O(N) elements but doing a simple write per, or are you doing O(N) full lookups on a hash table.

> So it seems we have to make a choice between optimizing for reads and optimizing for writes?

Yes

Reply:
> It appears that @TylerBrinkley's implementation also incurs O(n) cost for update operations.

Unless I'm mistaken I believe my implementation is O(n) only for `Insert` and `Remove` unless resizing is needed. See the chart here.

https://github.com/dotnet/runtime/issues/24826#issuecomment-362044463